### PR TITLE
[WIP] UCF101 prototype with utilities for video loading

### PR DESCRIPTION
**Pull request by bjuncek**
A simple `pyav` based set of utilities with a POC implementation for UCF101 dataset

### CODE REVIEW

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        self,
+        data: Tuple[np.ndarray, int],
+        *,
+        decoder: Optional[Callable[[io.IOBase], Dict[str, Any]]],
+    ) -> Dict[str, Any]:
+        annotations_d, file_d = data
+
+        label = annotations_d[1]
+        _path, file_handle = file_d
+        return {"path": _path, "file": file_handle, "target": label}
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:65]: We also need to handle the decoder here. By default, `decoder` will be `None`, but the user is free to pass something.

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+
+        annotations = resource_dps[0]
+        files = resource_dps[1]
+
+        annotations_dp = ZipArchiveReader(annotations)
+        annotations_dp = Filter(annotations_dp,
+                                self._filtername,
+                                fn_kwargs=dict(tgt=f"{config.split}list0{config.fold}.txt"))
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:87]:
```suggestion
        annotations_dp = Filter(annotations_dp, path_comparator("name", f"{config.split}list0{config.fold}.txt"))
```

You can import `path_comparator` from `torchvision.prototype.datasets.utils._internal`

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        annotations_dp = CSVParser(annotations_dp, delimiter=" ")
+        # COMMENT OUT FOR TESTING
+        annotations_dp = Shuffler(annotations_dp, buffer_size=INFINITE_BUFFER_SIZE)
+
+        files_dp = RarArchiveReader(files)
+        dp = KeyZipper(annotations_dp, files_dp, self._getname, self._getname)
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:93]: If you only pass a single key function to `KeyZipper`, it will be used for both datapipes. 

```suggestion
        dp = KeyZipper(annotations_dp, files_dp, path_accessor("name"))
```

You can import `path_accessor` from `torchvision.prototype.datasets.utils._internal`.

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            valid_options={'split': ["train", "test"], 'fold': ["1", "2", "3"]},
+            # categories=HERE / "ucf101.categories",
+            homepage="https://www.crcv.ucf.edu/data/UCF101.php",
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                "https://www.crcv.ucf.edu/data/UCF101/UCF101TrainTestSplits-RecognitionTask.zip",
+                sha256="",
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:47]:
```suggestion
                sha256="5c0d1a53b8ed364a2ac830a73f405e51bece7d98ce1254fd19ed4a36b224bd27",
```

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            ),
+            HttpResource(
+                "https://www.crcv.ucf.edu/data/UCF101/UCF101.rar",
+                sha256="",
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:51]:
```suggestion
                sha256="ca8dfadb4c891cb11316f94d52b6b0ac2a11994e67a0cae227180cd160bd8e55",
```

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            )
+        ]
+
+    def _collate_and_decode(
+        self,
+        data: Tuple[np.ndarray, int],
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:57]: This doesn't seem right. Is that an artifact from copy pasting from another dataset?

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+        annotations = resource_dps[0]
+        files = resource_dps[1]
+
+        annotations_dp = ZipArchiveReader(annotations)
+        annotations_dp = Filter(annotations_dp,
+                                self._filtername,
+                                fn_kwargs=dict(tgt=f"{config.split}list0{config.fold}.txt"))
+        annotations_dp = CSVParser(annotations_dp, delimiter=" ")
+        # COMMENT OUT FOR TESTING
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:89]: True, but should be removed before merge.

File: torchvision/prototype/datasets/utils/_dataset.py [hunk 8]
```diff
@@ -28,6 +28,7 @@
 class DatasetType(enum.Enum):
     RAW = enum.auto()
     IMAGE = enum.auto()
+    VIDEO = enum.auto()
```
Comment (pmeier) [inline torchvision/prototype/datasets/utils/_dataset.py:31]: We also need to update 

https://github.com/pytorch/vision/blob/0817f7f5983b05fdc34e06476ca270e2168c57b8/torchvision/prototype/datasets/_api.py#L54-L57

File: torchvision/prototype/datasets/decoder.py [hunk 9]
```diff
@@ -13,3 +16,17 @@ def raw(buffer: io.IOBase) -> torch.Tensor:
 
 def pil(buffer: io.IOBase, mode: str = "RGB") -> torch.Tensor:
     return pil_to_tensor(PIL.Image.open(buffer).convert(mode.upper()))
+
+
+def av_kf(buffer: io.IOBase, **read_video_kwargs: Any) -> Dict[str, Any]:
+    with unittest.mock.patch("torchvision.io.video.os.path.exists", return_value=True):
```
Comment (pmeier) [inline torchvision/prototype/datasets/decoder.py:22]: I don't think we need that anymore. I only added it in #4541 because I used `torchvision.io.video.read_video` and it enforces that the input is a file on disk although it can very well work with file handles.

File: torchvision/prototype/datasets/decoder.py [hunk 10]
```diff
@@ -13,3 +16,17 @@ def raw(buffer: io.IOBase) -> torch.Tensor:
 
 def pil(buffer: io.IOBase, mode: str = "RGB") -> torch.Tensor:
     return pil_to_tensor(PIL.Image.open(buffer).convert(mode.upper()))
+
+
+def av_kf(buffer: io.IOBase, **read_video_kwargs: Any) -> Dict[str, Any]:
```
Comment (pmeier) [inline torchvision/prototype/datasets/decoder.py:21]: I feel like a decoder in this file should follow the "one file handle, one decoded output" paradigm. That means, I would prefer to have an `av` function here that decodes the whole video. Given that you also don't use this function in the video utilities that you wrote, I think that should not be a problem.

File: torchvision/prototype/datasets/video_utils.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    def __init__(self, video_dp: IterDataPipe[Dict[str, Any]], transform=None) -> None:
+        """TorchData Iterdatapype that takes in video datapipe
+        and yields all keyframes in a video
+
+        Args:
+            video_dp (IterDataPipe[Dict[str, Any]]): Video dataset IterDataPipe
+            transform ([type], optional): Transform to be applied to every frame. Defaults to None.
+        """
+        self.datapipe = video_dp
+        self.transform = transform
```
Comment (pmeier) [inline torchvision/prototype/datasets/video_utils.py:19]: That is an anti-pattern we want to avoid. Transformations can applied in a separate `.map` call.

File: torchvision/prototype/datasets/video_utils.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+    def __iter__(self) -> Iterator[Dict[str, Any]]:
+        for video_d in self.datapipe:
+            buffer = video_d["file"]
```
Comment (pmeier) [inline torchvision/prototype/datasets/video_utils.py:23]: The key should probably be a parameter to the datapipe, because we probably won't enforce a fixed key so we can hardcode it here.
```suggestion
            buffer = video_d.pop("file")
```

File: torchvision/prototype/datasets/video_utils.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            with av.open(buffer, metadata_errors="ignore") as container:
+                stream = container.streams.video[0]
+                stream.codec_context.skip_frame = 'NONKEY'
+                for frame in container.decode(stream):
+                    img = frame.to_image()
+                    if self.transform is not None:
+                        img = self.transform(img)
+                    yield {
```
Comment (pmeier) [inline torchvision/prototype/datasets/video_utils.py:31]: We should make sure that we include all the information that we had previously and only replace `"file"` with new data. The easiest way to do that is to `pop` the key and then use the old data as base for the new dictionary.

```suggestion
                    yield dict(video_d, frame=img, ...)
```

**Review by pmeier — COMMENTED** [review]
Thanks a lot @bjuncek. I have some comments inline about the general infrastructure. I can't really comment on the validity of the video utility datapipes that you added, because I have to little experience with videos. I'll leave that up to other reviewers.

File: main.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+from torchvision.prototype import datasets
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply main.py:1]: Yes, exactly - this is to play around with and modify different params to mock the behaviour

File: torchvision/prototype/datasets/decoder.py (hunk 10 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/decoder.py:21]: actually, after thinking about it, all decoding is atm happening in the `video_utils` file so I'll revert this file since we don't use it

File: torchvision/prototype/datasets/video_utils.py (hunk 11 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/video_utils.py:19]: will remove it - force of habit I suppose :)

File: torchvision/prototype/datasets/decoder.py (hunk 10 above)
Comment (pmeier) [reply torchvision/prototype/datasets/decoder.py]: Fine by me. We can always add it later if the need arises.

File: torchvision/prototype/datasets/_builtin/ucf101.py (hunk 7 above)
Comment (fmassa) [reply torchvision/prototype/datasets/_builtin/ucf101.py:86]: I would really want our datasets to be deterministic outside of a DataLoader though. Making it stochastic will make it much harder to debug.

Maybe what we should do instead is have a new `DataLoaderOnlyShuffler` which is a no-op in general, and inside the dataloader it can be activated if `shuffle` is true.

Thoughts?

cc @VitalyFedyunin @ejguan

File: torchvision/prototype/datasets/_builtin/ucf101.py (hunk 6 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/_builtin/ucf101.py]: ah, yes; mea culpa - should be fixed

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    Dataset,
+    DatasetConfig,
+    DatasetInfo,
+    HttpResource,
+    OnlineResource,
+    DatasetType,
+)
+
+
+class ucf101(Dataset):
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:27]:
```suggestion
class UCF101(Dataset):
```

File: torchvision/prototype/datasets/_builtin/ucf101.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    """This is a base datapipe that returns a file handler of the video.
+    What we want to do is implement either several decoder options or additional
+    datapipe extensions to make this work.
+    """
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/ucf101.py:31]: True, but probably shouldn't be here.

File: torchvision/prototype/datasets/_builtin/ucf101.py (hunk 7 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/ucf101.py:84]: See pytorch/pytorch#65756.

File: torchvision/prototype/datasets/utils/_internal.py [hunk 17]
```diff
@@ -176,6 +177,38 @@ def __iter__(self) -> Iterator[Tuple[str, io.IOBase]]:
             yield path, decompressor(file)
 
 
+class RarArchiveReader(IterDataPipe[Tuple[str, io.BufferedIOBase]]):
```
Comment (pmeier) [inline torchvision/prototype/datasets/utils/_internal.py:180]: This should be landed in `torchdata`. See pytorch/data#41.

**Review by pmeier — APPROVED** [review]
Some minor comments inline. I didn't review the video datapipes and leaving that to someone with domain knowledge. Otherwise LGTM when CI is green! Thanks Bruno!

**Review by pmeier — CHANGES_REQUESTED** [review]
Just changing review status so that we don't forget to remove the `main.py` file before merge.

File: torchvision/prototype/datasets/utils/_internal.py (hunk 17 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/utils/_internal.py:180]: it's not for now. should I wait until it is?

**Comment by bjuncek** [comment]
Ok, so I've tried doing a pass on this, trying to fix the decoder inconsistency we've been talking about offline. 

I don't understand datapipes well enough to understand why pop from a dict would fail or why I'd need to annotate variables in a datapipe. Everything since `8f57ee6` has nothing to do with the functionality of the UCF101 dataset and can be reverted if you want to do a clean pass and then merge into this.

File: torchvision/prototype/datasets/utils/_internal.py (hunk 17 above)
Comment (pmeier) [reply torchvision/prototype/datasets/utils/_internal.py:188]: Nope, let's move forward here. If the PR in `torchdata` is merged, we can remove it here.

File: torchvision/prototype/datasets/video_utils.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+                        "clip": vframes,
+                        "pts": clip_pts,
+                        "range": (start_pts, end_pts),
+                        "video_meta": {
+                            "time_base": float(stream.time_base),
+                            "guessed_fps": float(stream.guessed_rate),
+                        },
+                        "path": video_d["path"],
+                        "target": video_d["target"],
+                    }
```
Comment (fmassa) [inline torchvision/prototype/datasets/video_utils.py:178]: Why not just do the following:
- sample `m` start positions
- for every start position, read `k` frames
- yield the `k` frames at once, `m` times

File: torchvision/prototype/datasets/video_utils.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
+import random
+from typing import Any, Dict, Iterator
+
+import av
+import numpy as np
+import torch
+from torchdata.datapipes.iter import IterDataPipe
+from torchvision.io import video, _video_opt
```
Comment (fmassa) [inline torchvision/prototype/datasets/video_utils.py:8]: I'm not sure if I would use `_video_opt` in here.

**Comment by pmeier** [comment]
@bjuncek I've cleaned up the PR with recent changes and also removed all the decoder changes. We will handle this separate from this PR following #5075.

File: torchvision/prototype/datasets/video_utils.py (hunk 18 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/video_utils.py:178]: Unless I'm missing something, this is exactly what I do:
1. sample starting positions (line 132)
2. for every start position (line 134) read `k` frames (line 140)
3. yield the frames as a sample (line 168)

Are you suggesting to take the yield outside of the loop? If so, is there any benefit to this?

File: torchvision/prototype/datasets/video_utils.py (hunk 19 above)
Comment (bjuncek) [reply torchvision/prototype/datasets/video_utils.py:8]: Sure.
Any particular reason why not?
